modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+  }
+
+  /*
+    Remove an element from this node's bag. The order in which elements are removed
+    are not guaranteed to be the same order it has been inserted. If this node's
+    bag is empty, it will attempt to steal elements from bags of other nodes.
+  */
+  proc remove() : (bool, eltType) {
+    return getPrivatizedThis.bag.remove();
+  }
+
+  /*
+    Obtain the number of elements held in all bags across all nodes. This method
+    is best-effort and can be non-deterministic for concurrent updates across nodes,
+    and may miss elements or even count duplicates resulting from any concurrent
+    insertion or removal operations.
+  */
+  proc getSize() : int {
+    var sz : atomic int;
+    forall loc in targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      forall segmentIdx in 0..#here.maxTaskPar {
+        sz.add(instance.bag.segments[segmentIdx].nElems.read() : int);
+      }
+    }
+
+    return sz.read();
+  }
+
+  /*
+    Performs a lookup to determine if the requested element exists in this bag.
+    This method is best-effort and can be non-deterministic for concurrent
+    updates across nodes, and may miss elements resulting from any concurrent
+    insertion or removal operations.
+  */
+  proc contains(elt : eltType) : bool {
+    var foundElt : atomic bool;
+    forall elem in getPrivatizedThis {
+      if elem == elt {
+        foundElt.write(true);
+      }
+    }
+    return foundElt.read();
+  }
+
+  /*
+    Clear all bags across all nodes in a best-effort approach. Elements added or
+    moved around from concurrent additions or removals may be missed while clearing.
+  */
+  proc clear() {
+    var localThis = getPrivatizedThis;
+    coforall loc in localThis.targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      forall segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        if segment.acquireIfNonEmpty(STATUS_REMOVE) {
+          var block = segment.headBlock;
+          while block != nil {
+            var tmp = block;
+            block = block.next;
+            delete tmp;
+          }
+
+          segment.headBlock = nil;
+          segment.tailBlock = nil;
+          segment.nElems.write(0);
+          segment.releaseStatus();
+        }
+      }
+    }
+  }
+
+  /*
+    Triggers a more static approach to load balancing, fairly redistributing all
+    elements fairly for bags across nodes. The result will result in all segments
+    having roughly the same amount of elements.
+
+    **Note:** This method is very heavy-weight in that it should not be called too
+    often. Dynamic work stealing handles cases where there is a relatively fair
+    distribution across majority of nodes, but this should be called when you have
+    a severe imbalance, or when you have a smaller number of elements to balance.
+    Furthermore, while this operation is parallel-safe, it should be called in a
+    sequential context.

```

Comment:
I think - but am not sure - that what you mean by "it should be called in a sequential context" is that it should only be called from one task at a time - since it redistributes the entire data structure and uses on statements etc to handle data elsewhere. I don't think "should be called in a sequential context" means that for me. I'd probably just say "Furthermore, while this operation is parallel-safe, it's only really useful to call it from one task at a time, since that call will update the entire distributed data structure".